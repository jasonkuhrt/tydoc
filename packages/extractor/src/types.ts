export * as Doc from './extractor/doc'
