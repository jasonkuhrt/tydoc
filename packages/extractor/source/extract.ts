import makeDebug from 'debug'
import dedent from 'dedent'
import * as fs from 'fs-jetpack'
import * as lo from 'lodash'
import { isEmpty } from 'lodash'
import * as path from 'path'
import * as tsm from 'ts-morph'
import { PackageJson } from 'type-fest'
import * as Doc from './doc'
import { scan } from './layout'
import * as Package from './lib/package'
import {
  absolutify,
  assertFileExists,
  downloadPackage,
  getPackageMain,
  JsFilePathToTsDeclarationFilePath,
  pathToModulePath,
  readPackageJson,
} from './lib/package-helpers'
import {
  DiagnosticFilter,
  getDiscriminantPropertiesOfUnionMembers,
  getFirstDeclarationOrThrow,
  getGenericType,
  getProperties,
  getSourceFileModulePath,
} from './lib/ts-helpers'
import { getLocationKind, getNodeFromTypePreferingAlias, hasAlias, isCallable, isPrimitive } from './utils'

const debug = makeDebug('tydoc:extract')
const debugExport = makeDebug('tydoc:extract:export')
const debugVisible = makeDebug('tydoc:extract:visible')
const debugWarn = makeDebug('tydoc:warn')

export interface FromProjectParams {
  /**
   * Paths to modules in project, relative to source root or absolute.
   */
  entrypoints: string[]
  /**
   * Should Tydoc settings be read from from the package file.
   */
  readSettingsFromJSON?: boolean
  /**
   * Sometimes a source entrypoint is fronted by a facade module that allows
   * package consumers to do e.g. `import foo from "bar/toto"` _instead of_
   * `import foo from "bar/dist/toto". Use this mapping to force tydoc to view
   * the given source modules (the keys) at the given package path (the values).
   *
   * @example
   *
   * Given project layout:
   *
   * ```
   * /src/foo/bar/toto.ts
   * ```
   *
   * The setting:
   *
   * ```ts
   * sourceModuleToPackagePathMappings: {
   *    "foo/bar/toto": "toto"
   * }
   * ```
   *
   * Will cause the `toto` module to be documented as being available at path:
   *
   * ```ts
   * import some from "thing/toto"
   * ```
   */
  sourceModuleToPackagePathMappings?: Record<string, string>
  layout?: {
    /**
     * Should Tydoc halt when TypeScript diagnostics are raised?
     *
     * Pass true to halt on any diagnostic
     *
     * Pass false to ignore all diagnostics
     *
     * Pass an array of filters to ignore only certain diagnostics.
     *
     * @default true
     *
     * @remarks
     *
     * Typically your package should be error free before documentation is extracted for it. However there may be cases where you want to bypass this check in general or for specific kinds of type check errors. For example a @ts-expect-error that is erroring because there is no error.
     *
     */
    validateTypeScriptDiagnostics?: boolean | DiagnosticFilter[]

    /**
     * Should the projectDir and sourceMainModulePath be validated that they exist on disk?
     *
     * @default true
     */
    validateExists?: boolean
    /**
     * Absolute path to the source main entrypoint.
     *
     * @default A discovery attempt is made by running heuristics against a combination of package.json and tsconfig.json however it only covers common patterns, not all possible setups.
     */
    sourceMainModulePath?: string
    /**
     * Specify the root of the project.
     *
     * @default The current working directory
     * @remarks This is useful for tests to avoid having to mock process.cwd
     */
    projectDir?: string
    sourceDir?: string
    packageJson?: PackageJson
    tsMorphProject?: tsm.Project
  }
}

export type FromPublishedParams = {
  packageName: string
  packageVersion?: string
  project?: tsm.Project
  downloadDir?: string
}

export type ExtractedPackageData = {
  docs: Doc.Package
  metadata: Package.Metadata
}

// todo support extracting fromPublished with additional entrypoints
export async function fromPublished(options: FromPublishedParams): Promise<ExtractedPackageData> {
  const tmpDir = await fs.tmpDirAsync()
  const projectDir = options.downloadDir ?? tmpDir.cwd()

  await downloadPackage({
    name: options.packageName,
    version: options.packageVersion,
    downloadDir: projectDir,
  })

  const packageJson = readPackageJson(projectDir)

  if (!packageJson) {
    throw new Error(
      `The downloaded package at ${projectDir} was not valid. It is missing a package.json file.`
    )
  }

  const packageJsonMain = getPackageMain(packageJson)
  debug('found packageJsonMain %s', packageJsonMain)

  const entrypointPath = JsFilePathToTsDeclarationFilePath(path.join(projectDir, packageJsonMain))
  debug('found entrypointPath %s', entrypointPath)

  assertFileExists(
    entrypointPath,
    `The downloaded package at ${projectDir} is not valid because it did not include TypeScript declaration files (looked for one at ${entrypointPath}).`
  )

  const project = options.project ?? new tsm.Project()

  project.addSourceFilesAtPaths(entrypointPath)

  const manager = new Doc.Manager({
    projectDir: projectDir,
    sourceDir: projectDir,
    sourceMainModulePath: entrypointPath,
  })

  project.getSourceFiles().forEach((sf) => {
    fromModule(manager, sf)
  })

  const metadata = await Package.getMetadata(options.packageName)

  return {
    docs: manager.EDD,
    metadata,
  }
}

/**
 * Recursively extract docs from the given project starting from the exports of
 * the given list of entrypoint modules. Everything that is reachable from the
 * exports will be considered part of the API.
 */
export function fromProject(options: FromProjectParams): Doc.Package {
  const layout = scan(options.layout)

  const sourceFiles = layout.tsMorphPoject.getSourceFiles()

  if (isEmpty(sourceFiles)) {
    throw new Error('No source files found in project to document.')
  }

  debug(
    'found project source files:',
    sourceFiles.map((sf) => sf.getFilePath())
  )

  /**
   * Find the corresponding source files for the given entrypoints
   */

  const sourceFileEntrypoints = options.entrypoints.map((givenEntryPoint) => {
    let givenEntryPointAbs = absolutify(layout.sourceDir, givenEntryPoint)

    /**
     * If given entrypoint is a folder then infer that to mean looking for
     * an index within it (just like how node module resolution works).
     *
     * In case user is working with in-memory ts-morph project do not care if the directory
     * does not actually exist.
     */
    if (fs.exists(givenEntryPointAbs) === 'dir') {
      givenEntryPointAbs = path.join(givenEntryPointAbs, 'index.ts')
    }

    debug('givenEntryPointAbs is %s', givenEntryPointAbs)

    const sourceFileEntryPoint = sourceFiles.find((sf) => {
      return getSourceFileModulePath(sf) === pathToModulePath(givenEntryPointAbs)
    })

    if (!sourceFileEntryPoint) {
      throw new Error(
        `Given entrypoint not found in project: "${givenEntryPointAbs}". Source files were:\n\n${sourceFiles
          .map(getSourceFileModulePath)
          .join('\n')}`
      )
    }

    return sourceFileEntryPoint
  })

  /**
   * Setup manager
   */

  const managerSettings = {
    sourceDir: layout.sourceDir,
    projectDir: layout.projectDir,
    sourceMainModulePath: layout.sourceMainModulePath,
    sourceModuleToPackagePathMappings: options.sourceModuleToPackagePathMappings,
  }

  if (options.readSettingsFromJSON) {
    lo.merge(managerSettings, layout.tydocSettingsInPackage)
  } else {
    debug('reading user settings from package.json is disabled')
  }

  const manager = new Doc.Manager(managerSettings)

  /**
   * Get EDD
   */

  sourceFileEntrypoints.forEach((sf) => {
    fromModule(manager, sf)
  })

  return manager.EDD
}

/**
 * Recursively extract docs starting from exports of the given module.
 * Everything that is reachable will be considered.
 */
export function fromModule(manager: Doc.Manager, sourceFile: tsm.SourceFile): Doc.Package {
  const mod = Doc.modFromSourceFile(manager, sourceFile)

  for (const [exportName, exportedDeclarations] of sourceFile.getExportedDeclarations()) {
    const n = exportedDeclarations[0]

    if (!n) {
      console.warn(
        dedent`
          Skipping named export "${exportName}" in module at "${sourceFile.getFilePath()}" because there were no exported declarations for it.

          This should not normally happen. Please open an issue with your use-case/context so we can investigate the issue.
        `
      )
      continue
    }

    const t = n.getType()
    debugExport('start')
    debugExport('-> node kind is %s', n.getKindName())
    debugExport('-> type text is %j', n.getType().getText())

    // if the node is a type alias and its type cannot find its way back to the
    // type alias then we are forced to run type alias extraction logic here.
    // So far we know this happens in typeof cases.
    let doc
    if (tsm.Node.isTypeAliasDeclaration(n) && !hasAlias(t)) {
      debugExport('type alias pointing to type that cannot back reference to the type alias %s', n.getText())
      doc = manager.indexTypeAliasNode(n, () =>
        Doc.alias({
          name: n.getName(),
          raw: {
            nodeFullText: n.getFullText(),
            nodeText: n.getText(),
            typeText: t.getText(),
          },
          typeParameters: typeParametersDocsFromType(manager, t),
          // todo getTSDocFromNode()
          ...getTSDoc(manager, t),
          type: fromType(manager, t),
        })
      )
    } else {
      doc = fromType(manager, t)
    }

    if (exportName === 'default') {
      mod.mainExport = doc
      continue
    }

    mod.namedExports.push(
      Doc.expor({
        name: exportName,
        type: doc,
        node: n,
      })
    )
  }

  manager.EDD.modules.push(mod)
  return manager.EDD
}

export function fromType(manager: Doc.Manager, t: tsm.Type): Doc.Node {
  debugVisible('start')
  debugVisible('-> type text is %j', t.getText())
  const locationKind = getLocationKind(t)
  debugVisible('-> type location is %s', locationKind)

  if (locationKind === 'dep') {
    debugVisible('-> location is a dependency, stopping here')
    return Doc.unsupported(getRaw(t), 'We do not support extracting from dependencies.')
  }
  if (locationKind === 'typeScriptStandardLibrary') {
    // we handle arrays specially below
    if (!t.isArray()) {
      debugVisible('-> location is standard lib and not array, stopping here')
      return Doc.unsupported(getRaw(t), 'We do not support extracting from the standard library.')
    }
  }
  if (locationKind === 'unknown') {
    debugWarn(
      '-> location is unknown, stopping here to be safe (code needs to be updated to handle this case)'
    )
    return Doc.unsupported(getRaw(t), 'We do not support extracting from types whose loation is unknown.')
  }

  if (t.isArray()) {
    debugVisible('-> type is array')
    const innerType = t.getArrayElementTypeOrThrow()
    debugVisible('-> handle array inner type %s', innerType.getText())
    return manager.indexTypeIfApplicable(t, () =>
      extractAliasIfOne(manager, t, Doc.array(fromType(manager, innerType)))
    )
  }

  const tt = getGenericType(t)
  if (tt) {
    debugVisible('-> type is an instance of a generic type: %s', tt.getText())
    return Doc.genericInstance({
      target: fromType(manager, tt) as any,
      ...getRaw(t),
    })
  }

  if (manager.isIndexable(t)) {
    debugVisible('-> type is indexable')
    const fqtn = manager.getFQTN(t)
    if (manager.isIndexed(fqtn)) {
      debugVisible('-> type is being documented as link to the type index (aka. cache hit)')
      return Doc.typeIndexRef(fqtn)
    }
  } else {
    debugVisible('-> type is not indexable')
  }

  if (t.isLiteral()) {
    debugVisible('-> type is literal')
    return Doc.literal({
      name: t.getText(),
      base: t.getBaseTypeOfLiteralType().getText(),
    })
  }
  if (isPrimitive(t)) {
    debugVisible('-> type is primitive')
    return Doc.prim(t.getText())
  }

  if (t.isTuple()) {
    debugVisible('-> type is tuple')
    return Doc.unsupported(getRaw(t), 'Tuples not yet implemented')
  }
  // Place before object becuase objects are superset.
  // Place before interface becuase interfaces can be callable
  if (isCallable(t)) {
    debugVisible('-> type is callable')
    return manager.indexTypeIfApplicable(t, () =>
      extractAliasIfOne(
        manager,
        t,
        Doc.callable({
          props: propertyDocsFromType(manager, t),
          sigs: sigDocsFromType(manager, t),
          ...getRaw(t),
        })
      )
    )
  }
  if (t.isInterface()) {
    debugVisible('-> type is interface')
    const sym = t.getSymbolOrThrow()
    return manager.indexTypeIfApplicable(t, () =>
      Doc.inter({
        name: sym.getName(),
        typeParameters: typeParametersDocsFromType(manager, t),
        props: propertyDocsFromType(manager, t),
        ...getTSDoc(manager, t),
        ...getRaw(t),
      })
    )
  }

  // Place after callable check because objects are superset.
  if (t.isObject()) {
    debugVisible('-> type is object')
    return manager.indexTypeIfApplicable(t, () =>
      extractAliasIfOne(
        manager,
        t,
        Doc.obj({
          props: propertyDocsFromType(manager, t),
          ...getRaw(t),
        })
      )
    )
  }
  if (t.isUnion()) {
    debugVisible('-> type is union')
    return manager.indexTypeIfApplicable(t, () => {
      const members = t.getUnionTypes()
      const discriminantProperties = getDiscriminantPropertiesOfUnionMembers(members)
      return extractAliasIfOne(
        manager,
        t,
        Doc.union({
          ...getRaw(t),
          discriminantProperties: discriminantProperties.map((p) => p.getName()),
          types: members.map((tm) => {
            debugVisible('-> handle union member %s', tm.getText())
            // todo no extract alias here ...
            return extractAliasIfOne(manager, tm, fromType(manager, tm))
          }),
        })
      )
    })
  }
  if (t.isIntersection()) {
    debugVisible('-> type is intersection')
    return manager.indexTypeIfApplicable(t, () =>
      extractAliasIfOne(
        manager,
        t,
        Doc.intersection({
          ...getRaw(t),
          types: t.getIntersectionTypes().map((tm) => {
            debugVisible('-> handle intersection member %s', tm.getText())
            return fromType(manager, tm)
          }),
        })
      )
    )
  }
  debugWarn('unsupported kind of type %s', t.getText())
  return Doc.unsupported(getRaw(t), `This kind of type is not supported: "${t.getText()}"`)
}

/**
 * Get the docs for the type parameters of the given type
 */
function typeParametersDocsFromType(docs: Doc.Manager, t: tsm.Type): Doc.TypeParameter[] {
  // todo why does ts-moprh separate type arg access?
  const params = t.isInterface() ? t.getTypeArguments() : t.getAliasTypeArguments()
  return params.map((param) => {
    const sym = param.getSymbolOrThrow()
    const defaultType = param.getDefault()
    const defaultTypeDocs = defaultType ? fromType(docs, defaultType) : null

    // todo use a better function than fromType that is for extracting inline or indexed types
    if (defaultTypeDocs) {
      if (defaultTypeDocs.kind === 'interface') {
        throw new Error(`Type parameters cannot be an inline interface`)
      }
      if (defaultTypeDocs.kind === 'alias') {
        throw new Error(`Type parameters cannot be an inline alias`)
      }
    }

    return {
      name: sym.getName(),
      default: defaultTypeDocs,
      ...getRaw(param),
    }
  })
}

/**
 * Extract docs from a type's call signatures. There may be multiple because of
 * function overloading.
 *
 * Note not only functions and methods may have call signatures
 * but anything callable which means objects too (in JS a function is an object
 * and in TS this as modelled as object types being able to be made callable).
 */
function sigDocsFromType(docs: Doc.Manager, t: tsm.Type): Doc.DocSig[] {
  return t.getCallSignatures().map((sig) => {
    const tRet = sig.getReturnType()
    const params = sig.getParameters()
    debugVisible('handle callable return: %s', tRet.getText())
    return Doc.sig({
      return: fromType(docs, tRet),
      params: params.map((p) => {
        const n = getFirstDeclarationOrThrow(p)
        const paramName = p.getName()
        const paramType = n.getType()
        // what is this method for then? It was just returning an `any` type
        // const paramType = p.getDeclaredType()
        // prettier-ignore
        debugVisible('handle callable param: %s %s %s', n.getKindName(), paramName, paramType.getText())
        return Doc.sigParam({
          name: paramName,
          type: fromType(docs, paramType),
        })
      }),
    })
  })
}

/**
 * Extract docs from the type's properties.
 */
function propertyDocsFromType(docs: Doc.Manager, t: tsm.Type): Doc.DocProp[] {
  return getProperties(t).map((p) => {
    const propName = p.getName()
    const propType = p.getType()
    // what is this method for then? It was just returning an `any` type
    // const propType = p.getDeclaredType()
    debugVisible('handle property: %s %s %s', p.getKindName(), propName, propType.getText())
    // Do not try to index type here. Must come after index lookup.
    return Doc.prop({
      name: propName,
      type: fromType(docs, propType),
    })
  })
}

/**
 * Extract doc for the type alias pointing to the given type, if any. The
 * extracted doc node is then wrapped around the given doc node and returned.
 * Thus the given doc node may be returned as-is if no alias found or returned
 * wrapped in doc for the found alias.
 */
function extractAliasIfOne(manager: Doc.Manager, t: tsm.Type, doc: Doc.Node): Doc.Node {
  // is it possible to get alias of aliases? It seems the checker "compacts"
  // these and if we __really__ wanted to "see" the chain we'd have to go the
  // node AST way.
  // debug(as?.getAliasedSymbol()?.getName())
  if (!hasAlias(t)) {
    return doc
  }
  const asym = t.getAliasSymbol()!
  debug('-> type had alias %s (extracting a doc node for it)', asym.getName())
  return Doc.alias({
    name: asym.getName(),
    type: doc,
    typeParameters: typeParametersDocsFromType(manager, t),
    ...getRaw(t),
    ...getTSDoc(manager, t),
  })
}

/**
 * Get raw doc information for the given type.
 */
function getRaw(t: tsm.Type): Doc.RawFrag {
  /**
   * The use-alias-outside-current-scope flag makes it so that the type for
   * something imported is seen not as `import(...).<name>` but the actual type
   * from the thing's origin.
   */
  const typeText = t.getText(undefined, tsm.ts.TypeFormatFlags.UseAliasDefinedOutsideCurrentScope).trim()

  const node = getNodeFromTypePreferingAlias(t)

  if (!node) {
    return {
      raw: {
        typeText: typeText,
        // todo null instead of empty string?
        nodeFullText: '',
        nodeText: '',
      },
    }
  }

  return {
    raw: {
      typeText: typeText,
      nodeText: node.getText().trim(),
      nodeFullText: node.getFullText().trim(),
    },
  }
}

/**
 * Extract tsdoc docs for the given type.
 *
 * @return
 *
 * If the node associated with the type does not support JSDoc (not all node
 * types do) then `null` is returned.
 *
 * If the node associated with the type does not have JSDoc present in the
 * source code then `null` is returned.
 */
export function getTSDoc(manager: Doc.Manager, t: tsm.Type): Doc.TSDocFrag {
  const n = getNodeFromTypePreferingAlias(t)
  let docFragTSDoc: Doc.TSDocFrag['tsdoc']

  if (!n) {
    docFragTSDoc = null
  } else if (!tsm.Node.isJSDocableNode(n)) {
    docFragTSDoc = null
  } else {
    const jsDocNode = n.getJsDocs()[0]
    if (!jsDocNode) {
      docFragTSDoc = null
    } else {
      docFragTSDoc = Doc.tsDocFromText(manager, jsDocNode.getText())
    }
  }

  return {
    tsdoc: docFragTSDoc,
  }
}
