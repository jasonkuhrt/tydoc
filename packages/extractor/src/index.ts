export * from './api/extractor/extract'
export { fromPublished } from './api/extractor/fromPublished'
export { Options as RenderMarkdownOptions, render as renderMarkdown } from './api/renderers/markdown'
