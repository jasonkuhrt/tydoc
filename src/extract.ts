import * as tsm from 'ts-morph'
import { inspect } from 'util'
import { casesHandled } from './utils'

/**
 * The root of documentation data.
 */
export interface Docs {
  terms: DocItem[]
  types: DocItem[]
  hybrids: DocItem[]
  length: number
}

/**
 * Create a new set of docs.
 */
function createDocs(): Docs {
  return {
    terms: [],
    hybrids: [],
    types: [],
    length: 0,
  }
}

/**
 * Add a doc item to docs.
 */
function addDoc(docs: Docs, doc: DocItem): Docs {
  docs[
    doc.languageLevel === 'term'
      ? 'terms'
      : doc.languageLevel === 'type'
      ? 'types'
      : doc.languageLevel === 'hybrid'
      ? 'hybrids'
      : casesHandled(doc.languageLevel)
  ].push(doc)
  docs.length++
  return docs
}

interface JSDocBlock {
  source: string
}

/**
 * It is possible for multiple jsDoc blocks to appear in succession. When
 * source code is authored that way, the jsDoc blocks _after_ the one closest
 * to the code item appear here. For example:
 *
 *    /**
 *     *   foo 2           <-- Goes into additional jsDoc blocks group
 *     *\/
 *    /**
 *     *   foo 1           <-- Considered the primary jsDoc block
 *     *\/
 *     const a = 1
 *
 */
interface JSDocContent {
  /**
   * The jsDoc block closest to the code.
   */
  primary: JSDocBlock
  additional: JSDocBlock[]
}

interface TypeData {
  name: string
}

interface DocBase {
  name: string
  jsDoc: null | JSDocContent
  text: string
  /**
   * This is about if the item is from the JavaScript language or the TypeScript
   * type system. "term" refers to JavaScript values. "type" refers to
   * TypeSript types. "hybrid" refers to constructs span both levels, such as classes.
   */
  languageLevel: 'term' | 'type' | 'hybrid'
  sourceLocation: {
    filePath: string
    fileLine: number
  }
}

export interface DocFunction extends DocBase {
  kind: 'function'
  signature: {
    parameters: { name: string; type: TypeData }[]
    return: TypeData
  }
}

export interface DocVariable extends DocBase {
  kind: 'variable'
  type: TypeData
}

export interface DocTypeAlias extends DocBase {
  kind: 'typeAlias'
}

export type DocItem = DocFunction | DocVariable | DocTypeAlias

/**
 * Extract docs from the module found at the given file path.
 */
export function extractDocsFromModuleAtPath(filePath: string) {
  const project = new tsm.Project({ addFilesFromTsConfig: true })

  project.addSourceFileAtPathIfExists(filePath)

  const source = project.getSourceFile(filePath)

  if (!source) {
    throw new Error(`No file at ${filePath}`)
  }

  return extractDocsFromModule(source)
}

/**
 * Extract docs from the given module.
 */
export function extractDocsFromModule(sourceFile: tsm.SourceFile): Docs {
  const exs = sourceFile.getExportedDeclarations()
  const docs = createDocs()

  for (const [name, declarations] of exs) {
    const doc = extractDocsFromDeclaration(name, declarations[0])
    addDoc(docs, doc)
  }

  return docs
}

/**
 * Extracts docs from the given declaration.
 */
function extractDocsFromDeclaration(
  name: string,
  dec: tsm.ExportedDeclarations
): DocItem {
  if (dec instanceof tsm.FunctionDeclaration) {
    return {
      kind: 'function',
      languageLevel: 'term',
      name,
      text: dec.getText(false),
      signature: {
        parameters: dec.getParameters().map(
          // p => `name: ${p.getName()} - type:
          // ${extractTypeData(p.getType())}`
          p => ({ name: p.getName(), type: getTypeData(p.getType()) })
        ),
        return: getTypeData(dec.getReturnType()),
      },
      jsDoc: getJSDocContent(dec),
      ...extractCommonDocDataFromDeclaration(dec),
    }
  }

  if (dec instanceof tsm.VariableDeclaration) {
    // A variable declaration is within a variable declaration list is inside a
    // variable statement. JSDoc lives at the variable statement level.
    const ctx = dec.getParent().getParent()
    if (ctx instanceof tsm.VariableStatement) {
      // If the variable is pointing to a function we will treat it like as if
      // it were a function declaration.
      const initializer = dec.getInitializer()
      if (
        initializer &&
        (initializer instanceof tsm.ArrowFunction ||
          initializer instanceof tsm.FunctionExpression)
      ) {
        const docs = extractDocsFromFunction(initializer) as DocFunction
        docs.jsDoc = getJSDocContent(ctx)
        docs.name = docs.name || name
        return docs
      }

      // const type = initializer && initializer.getType()
      const typeName = dec.getType().getText()
      return {
        kind: 'variable',
        languageLevel: 'term',
        name,
        // type: type && extractTypeData(type),
        type: {
          name: typeName,
        },
        text: dec.getText(false),
        jsDoc: getJSDocContent(ctx),
        ...extractCommonDocDataFromDeclaration(dec),
      }
    }
  }

  if (dec instanceof tsm.TypeAliasDeclaration) {
    return {
      kind: 'typeAlias',
      languageLevel: 'type',
      name,
      text: dec.getText(false),
      jsDoc: getJSDocContent(dec),
      ...extractCommonDocDataFromDeclaration(dec),
    }
  }

  // todo
  // casesHandled(declaration)

  throw new Error(
    `unknown kind of declaration or declaration scenario\n\n${inspect(dec)}`
  )
}

/**
 * Extract doc data that is common to all doc items from the given declaration.
 */
function extractCommonDocDataFromDeclaration(dec: tsm.ExportedDeclarations) {
  const sourceFile = dec.getSourceFile()
  const filePath = sourceFile.getFilePath()
  const fileLine = dec.getStartLineNumber()
  const commonDocData = {
    sourceLocation: {
      filePath,
      fileLine,
    },
  }
  return commonDocData
}

/**
 * Extract docs from the given Function like node. This does not quite return
 * full doc data because jsDoc is kept on variable declarations and name should
 * be the exported one not the maybe-present explicit function expression name.
 */
function extractDocsFromFunction(
  node: tsm.ArrowFunction | tsm.FunctionExpression
): Omit<DocFunction, 'name' | 'jsDoc'> {
  // todo anything useful we can by revealing explicitly named function expressions?
  // let maybeName = ''
  // if (dec instanceof tsm.FunctionExpression) {
  //   maybeName = dec.getName() || ''
  // }
  return {
    kind: 'function',
    languageLevel: 'term',
    text: node.getText(false),
    signature: {
      parameters: node
        .getParameters()
        .map(p => ({ name: p.getName(), type: getTypeData(p.getType()) })),
      return: getTypeData(node.getReturnType()),
    },
    ...extractCommonDocDataFromDeclaration(node),
  }
}

/**
 * Extract JSDoc from node. Tease apart primary block from additional blocks.
 */
function getJSDocContent(node: tsm.JSDocableNode): null | JSDocContent {
  const jsDocs = node.getJsDocs().map(doc => ({ source: doc.getFullText() }))

  if (jsDocs.length === 0) return null
  if (jsDocs.length === 1) return { primary: jsDocs[0], additional: [] }

  return { primary: jsDocs.pop()!, additional: jsDocs }
}

function getTypeData(type: tsm.Type): TypeData {
  return {
    name: type.getText(),
  }
}

// type TypeMetaData = {
//   text: string
//   properties: {
//     name: string
//     type: any // todo TypeMetaData
//   }[]
// }

// function extractTypeData(type: tsm.Type<tsm.ts.Type>): TypeMetaData {
//   const properties = type.getApparentProperties()
//   return {
//     text: type.getText(),
//     properties: properties.map(p => ({
//       name: p.getName(),
//       type: extractTypeData(p.getDeclaredType()),
//     })),
//   }
// }
