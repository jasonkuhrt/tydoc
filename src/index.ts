import * as tsm from 'ts-morph'
import { inspect } from 'util'

export function extractDocsAndTypesFromModuleAtPath(filePath: string) {
  const project = new tsm.Project({ addFilesFromTsConfig: true })

  project.addSourceFileAtPathIfExists(filePath)

  const source = project.getSourceFile(filePath)

  if (!source) {
    throw new Error(`No file at ${filePath}`)
  }

  return extractDocsAndTypesFromSourceFile(source)
}

function extractDocsAndTypesFromSourceFile(sourceFile: tsm.SourceFile) {
  const exs = sourceFile.getExportedDeclarations()

  const docs = []
  for (const [name, declarations] of exs) {
    docs.push(extractDocsAndTypesFromDeclaration(name, declarations[0]))
  }

  return docs
}

interface JSDocBlock {
  source: string
}
/**
 * It is possible for multiple jsDoc blocks to appear in succession. When
 * source code is authored that way, the jsDoc blocks _after_ the one closest
 * to the code item appear here. For example:
 *
 *    /**
 *     *   foo 2           <-- Goes into additional jsDoc blocks group
 *     *\/
 *    /**
 *     *   foo 1           <-- Considered the primary jsDoc block
 *     *\/
 *     const a = 1
 *
 */
interface JSDocContent {
  /**
   * The jsDoc block closest to the code.
   */
  primary: JSDocBlock
  additional: JSDocBlock[]
}

interface DocBase {
  name: string
  jsDoc: null | JSDocContent
  location: {
    filePath: string
    fileLine: number
  }
}

interface DocFunction extends DocBase {
  kind: 'function'
  signature: {
    parameters: any[]
    return: any
  }
}

interface DocVariable extends DocBase {
  kind: 'variable'
  type: any
}

type DocItem = DocFunction | DocVariable

function extractDocsAndTypesFromDeclaration(
  name: string,
  declaration: tsm.ExportedDeclarations
): DocItem {
  const sourceFile = declaration.getSourceFile()
  const filePath = sourceFile.getFilePath()
  const fileLine = declaration.getStartLineNumber()

  if (declaration instanceof tsm.FunctionDeclaration) {
    return {
      kind: 'function',
      name,
      signature: {
        parameters: declaration
          .getParameters()
          .map(
            p => `name: ${p.getName()} - type: ${extractTypeData(p.getType())}`
          ),
        return: declaration.getReturnType().getText(),
      },
      jsDoc: getJSDocContent(declaration),
      location: {
        filePath,
        fileLine,
      },
    }
  }

  if (declaration instanceof tsm.VariableDeclaration) {
    // A variable declaration is within a variable declaration list is inside a
    // variable statement. JSDoc lives at the variable statement level.
    const ctx = declaration.getParent().getParent()
    const initializer = declaration.getInitializer()
    if (ctx instanceof tsm.VariableStatement) {
      // todo when would this be undefined?
      const type = initializer && initializer.getType()
      return {
        kind: 'variable',
        name,
        type: type && extractTypeData(type),
        jsDoc: getJSDocContent(ctx),
        location: {
          filePath,
          fileLine,
        },
      }
    }
  }

  throw new Error(
    `unknown kind of declaration or declaration scenario ${inspect(
      declaration
    )}`
  )
}

/**
 * Extract JSDoc from node. Tease apart primary block from additional blocks.
 */
function getJSDocContent(node: tsm.JSDocableNode): null | JSDocContent {
  const jsDocs = node.getJsDocs().map(doc => ({ source: doc.getFullText() }))

  if (jsDocs.length === 0) return null
  if (jsDocs.length === 1) return { primary: jsDocs[0], additional: [] }

  return { primary: jsDocs.pop()!, additional: jsDocs }
}

type TypeMetaData = {
  text: string
  properties: {
    name: string
    type: any // todo TypeMetaData
  }[]
}

function extractTypeData(type: tsm.Type<tsm.ts.Type>): TypeMetaData {
  const properties = type.getApparentProperties()

  return {
    text: type.getText(),
    properties: properties.map(p => ({
      name: p.getName(),
      type: extractTypeData(p.getDeclaredType()),
    })),
  }
}

// const docs = extractDocsAndTypesFromModuleAtPath(
//   '/Users/jasonkuhrt/projects/nexus/nexus-future/src/index.ts'
// )

// docs //?+
// inspect(docs, { depth: null }) //?
