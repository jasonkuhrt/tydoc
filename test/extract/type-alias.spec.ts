import '../../src/extract'
const ctx = createContext()

it('extracts doc from type aliases', () => {
  expect(
    ctx.given(`
      export type Foo = {}
      /**
       * boo
       */
      export type Bar = {}
    `)
  ).toMatchSnapshot()
})
