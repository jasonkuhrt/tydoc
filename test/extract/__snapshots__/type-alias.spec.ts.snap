// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extracts doc from type aliases 1`] = `
Object {
  "hybrids": Array [],
  "length": 2,
  "terms": Array [],
  "types": Array [
    Object {
      "jsDoc": null,
      "kind": "typeAlias",
      "languageLevel": "type",
      "name": "Foo",
      "properties": Array [],
      "sourceLocation": Object {
        "fileLine": 1,
        "filePath": "/test.ts",
      },
      "text": "export type Foo = {};",
      "textWithJSDoc": "export type Foo = {};",
    },
    Object {
      "jsDoc": Object {
        "additional": Array [],
        "primary": Object {
          "source": "boo",
        },
      },
      "kind": "typeAlias",
      "languageLevel": "type",
      "name": "Bar",
      "properties": Array [
        Object {
          "jsDoc": Object {
            "additional": Array [
              Object {
                "source": "Doc about a 2",
              },
            ],
            "primary": Object {
              "source": "Doc about a 1",
            },
          },
          "name": "a",
          "type": Object {
            "isPrimitive": true,
            "name": "string",
          },
        },
        Object {
          "jsDoc": Object {
            "additional": Array [],
            "primary": Object {
              "source": "Doc about b",
            },
          },
          "name": "b",
          "type": Object {
            "isPrimitive": true,
            "name": "number",
          },
        },
      ],
      "sourceLocation": Object {
        "fileLine": 6,
        "filePath": "/test.ts",
      },
      "text": "export type Bar = {
  /**
   * Doc about a 2
   */
  /**
   * Doc about a 1
   */
  a: string;
  /**
   * Doc about b
   */
  b: number;
};",
      "textWithJSDoc": "/**
 * boo
 */
export type Bar = {
  /**
   * Doc about a 2
   */
  /**
   * Doc about a 1
   */
  a: string;
  /**
   * Doc about b
   */
  b: number;
};",
    },
  ],
}
`;
